radixdlt/src/main/java/com/radixdlt/consensus/sync/SyncManager.java

```diff
@@ -0,0 +1,150 @@
+	private long targetVersion = -1;
+	private ScheduledFuture<?> timeoutChecker;
+	private LongConsumer onTarget = target -> { };
+	private LongConsumer onVersion = version -> { };
+
+	public SyncManager(Consumer<CommittedAtom> atomProcessor, LongSupplier versionProvider, int batchSize,
+			long patience) {
+		this.atomProcessor = atomProcessor;
+		this.versionProvider = versionProvider;
+		this.batchSize = batchSize;
+		this.patience = patience;
+		this.maxAtomsQueueSize = MAX_REQUESTS_TO_SEND * batchSize * 2;
+	}
+
+	public void syncToVersion(long targetVersion, LongConsumer requestSender) {
+		executorService.execute(() -> {
+			if (targetVersion <= this.targetVersion) {
+				return;
+			}
+			long crtVersion = versionProvider.getAsLong();
+			if (crtVersion >= targetVersion) {
+				return;
+			}
+			this.targetVersion = targetVersion;
+			this.requestSender = requestSender;
+			sendSyncRequests();
+		});
+	}
+
+	public void syncAtoms(List<CommittedAtom> atoms) {
+		executorService.execute(() -> {
+			long crtVersion = versionProvider.getAsLong();
+			for (CommittedAtom atom : atoms) {
+				long atomVersion = atom.getVertexMetadata().getStateVersion();
+				if (atomVersion > crtVersion) {
+					if (commitedAtoms.size() < maxAtomsQueueSize) {
+						commitedAtoms.add(atom);
+					} else {
+						CommittedAtom last = commitedAtoms.last();
+						if (last.getVertexMetadata().getStateVersion() > atomVersion) {
+							commitedAtoms.pollLast();
+							commitedAtoms.add(atom);
+						}
+					}
+				}
+			}
+			applyAtoms();
+		});
+	}
+
+	private void applyAtoms() {
+		long initialVersion = versionProvider.getAsLong();
+		Iterator<CommittedAtom> it = commitedAtoms.iterator();
+		while (it.hasNext()) {
+			CommittedAtom crtAtom = it.next();
+			long atomVersion = crtAtom.getVertexMetadata().getStateVersion();
+			if (atomVersion <= versionProvider.getAsLong()) {
+				it.remove();
+			} else if (atomVersion == versionProvider.getAsLong() + 1) {
+				atomProcessor.accept(crtAtom);
+				it.remove();
+			} else {
+				break;
+			}
+		}
+
+		long newVersion = versionProvider.getAsLong();
+		if (newVersion > initialVersion) {
+			onVersion.accept(newVersion);
+			if (versionProvider.getAsLong() >= targetVersion) {
+				if (timeoutChecker != null) {
+					timeoutChecker.cancel(false);
+				}
+				if (initialVersion < targetVersion) {
+					onTarget.accept(targetVersion);
+				}
+			}
+		}
+	}
+
+	private void sendSyncRequests() {
+		long crtVersion = versionProvider.getAsLong();
+		if (crtVersion >= targetVersion) {
+			return;
+		}
+		long size = ((targetVersion - crtVersion) / batchSize);
+		if ((targetVersion - crtVersion) % batchSize > 0) {
+			size += 1;
+		}
+		size = Math.min(size, MAX_REQUESTS_TO_SEND);
+		for (long i = 0; i < size; i++) {
+			requestSender.accept(crtVersion + batchSize * i);
+		}
+		timeoutChecker = executorService.schedule(this::sendSyncRequests, patience, TimeUnit.SECONDS);
+	}
+
+	public long getTargetVersion() {
```

Comment:
These three methods should be made package-local and annotated with VisibleForTesting

Reply:
Ok. I thought that may be useful in the future, but for now let's reduce there visibility.